File: Assignment 6/Dilsin_Amirtharajah_Assignment6.cpp
#include <iostream> 
#include <string>   
#include <sstream>  
#include <algorithm> 
#include <vector>
#include <climits>  
using namespace std; 

struct Entry { // structure to store the input
    int player_id;    
    int player_score; 
}; 

class high_score_class { 
public:              
    Entry* data;     // pointer to entry
    int size;
    int capacity;

    high_score_class() { // default 
        capacity = 4;                
        data = new Entry[capacity];  
        size = 0;                    
    }                                

    
    ~high_score_class() { // destructor 
        delete[] data; // delete array
    }                  

    
    high_score_class(const high_score_class& other) { // copy constructor 
        capacity = other.capacity;            
        size = other.size;                    
        data = new Entry[capacity];           // copy and new array
        for (int i = 0; i < size; ++i)
            data[i] = other.data[i]; 
    }                                


    high_score_class& operator=(const high_score_class& other) { 
        if (this != &other) {                            
            delete[] data;                               
            capacity = other.capacity;                   // copy capacity
            size = other.size;                           // copy size
            data = new Entry[capacity];                  
            for (int i = 0; i < size; ++i)
                data[i] = other.data[i]; 
        }                                
        return *this;                    
    }

    
    int findIndexById(int player_id) const { // search function 
        for (int i = 0; i < size; ++i) { 
            if (data[i].player_id == player_id)
                return i; 
        }                 
        return -1;        
    }                     

    
    void checking_capacity_size() {             
        if (size < capacity) return;    
        int newCap = capacity * 2;      // double capacity
        Entry* newData = new Entry[newCap]; 
        for (int i = 0; i < size; ++i)
            newData[i] = data[i]; 
        delete[] data;            
        data = newData;           
        capacity = newCap;        
    }                             

    
    void addEntry(int player_id, int player_score) { // add new entry
        checking_capacity_size();              // checking capacity
        data[size].player_id = player_id;            
        data[size].player_score = player_score;      
        ++size;                        
    }                                  


    bool removeById(int player_id) { 
        int idx = findIndexById(player_id); 
        if (idx == -1) return false; 
        for (int i = idx; i + 1 < size; ++i)
            data[i] = data[i + 1]; 
        --size;
        return true;               
    }                              

   
    void updateByPointer(Entry* e, int newScore) { // update using pointer
        if (e) e->player_score = newScore;                
    }                                              

    
    void updateByReference(Entry& e, int newScore) { // update using reference
        e.player_score = newScore;
    }                                                


    void boostByPointer(Entry* e, int delta) { // boost score
        if (!e) return;                        
        long long temp = (long long)e->player_score + (long long)delta; 
        if (temp < 0) temp = 0;               // set min
        if (temp > 1000000) temp = 1000000;   // set max
        e->player_score = (int)temp;
    }

    void boostByReference(Entry& e, int delta) { // boost score by reference
        long long temp = (long long)e.player_score + (long long)delta; 
        if (temp < 0) temp = 0;                                 // set min
        if (temp > 1000000) temp = 1000000;                     // set max
        e.player_score = (int)temp;                                    
    }                                                           
}; 


pair<bool, int> try_to_parse_int(const string& s) { // parse string
    stringstream ss(s);                        
    long long tmp;                             
    if (!(ss >> tmp)) return {false, 0};       // failed 

    char c;
    if (ss >> c) return {false, 0};            // invalid if theres others
    
    if (tmp < INT_MIN || tmp > INT_MAX) return {false, 0}; 
    return {true, (int)tmp};
} 


void printMenu() { 
    cout << "High player_score Lab" << endl;
    cout << "1. Add Entry" << endl;
    cout << "2. View Entry" << endl;
    cout << "3. Update player_score" << endl;
    cout << "4. Remove Entry" << endl;
    cout << "5. Leaderboard (Top K)" << endl;
    cout << "6. Stats" << endl;
    cout << "7. Pointer Tools" << endl;
    cout << "8. Exit" << endl;
    cout << "Enter your choice: ";
}


int main() {
    high_score_class lab;
    string line;

    while (true) {
        printMenu();
        if (!getline(cin, line)) return 0;
        auto parsed = try_to_parse_int(line);
        if (!parsed.first) {
            cout << "Invalid input." << endl << endl;
            continue;
        }
        int choice = parsed.second;

        if (choice < 1 || choice > 8) {
            cout << "Invalid choice." << endl << endl;
            continue;
        }

        if (choice == 8) {
            cout << "Goodbye!" << endl;
            break;
        }

        if (choice == 1) {
            int idVal = 0;
            while (true) {
                cout << "Enter player_id: ";
                if (!getline(cin, line)) return 0;
                auto p = try_to_parse_int(line);
                if (!p.first) {
                    cout << "Invalid input." << endl;
                    continue;
                }
                idVal = p.second;
                if (idVal <= 0) {
                    cout << "Error: player_id must be greater than 0." << endl;
                    continue;
                }
                if (lab.findIndexById(idVal) != -1) {
                    cout << "Error: player_id already exists." << endl;
                    continue;
                }
                break;
            }

            int scoreVal = 0;
            while (true) {
                cout << "Enter player_score: ";
                if (!getline(cin, line)) return 0;
                auto p = try_to_parse_int(line);
                if (!p.first) {
                    cout << "Invalid input." << endl;
                    continue;
                }
                scoreVal = p.second;
                if (scoreVal < 0 || scoreVal > 1000000) {
                    cout << "Error: player_score out of range." << endl;
                    continue;
                }
                break;
            }

            lab.addEntry(idVal, scoreVal);
            cout << "Entry added!" << endl << endl;
            continue;
        }

        if (choice == 2) {
            cout << "Enter player_id: ";
            if (!getline(cin, line)) return 0;
            auto p = try_to_parse_int(line);
            if (!p.first) {
                cout << "Invalid input." << endl << endl;
                continue;
            }
            int idVal = p.second;
            int idx = lab.findIndexById(idVal);
            if (idx == -1) {
                cout << "Error: player_id not found." << endl << endl;
                continue;
            }
            cout << "player_id: " << lab.data[idx].player_id << " | player_score: " << lab.data[idx].player_score << endl << endl;
            continue;
        }

        if (choice == 5) { 
            cout << "Enter K: ";
            if (!getline(cin, line)) return 0;
            auto p = try_to_parse_int(line);
            if (!p.first) {
                cout << "Invalid input." << endl << endl;
                continue;
            }
            int K = p.second;
            if (K <= 0) {
                cout << "Error: K must be greater than 0." << endl << endl;
                continue;
            }
            if (lab.size == 0) {
                cout << endl;
                continue;
            }

            vector<Entry> temp;
            for (int i = 0; i < lab.size; ++i)
                temp.push_back(lab.data[i]);

            sort(temp.begin(), temp.end(), [](const Entry& a, const Entry& b) {
                if (a.player_score != b.player_score)
                    return a.player_score > b.player_score;
                return a.player_id < b.player_id;
            });

            
            int toShow = std::min<int>((int)temp.size(), K);
            for (int i = 0; i < toShow; ++i)
                cout << (i + 1) << ". player_id: " << temp[i].player_id << " | player_score: " << temp[i].player_score << endl;

            cout << endl;
            continue;
        }
    }
    return 0;
}

File: Assignment 6/Dilsin_Amirtharajah_Assignment6.py
capacity = 4  # initial capacity 
data_version_one = [None] * capacity  
size = 0  # entry starting

def try_to_parse_int(s):  # to parse string
    try:  
        val = int(s.strip())
        return True, val  # removing whitespace and return
    except:  
        return False, 0  # failed

# to find index
def find_index_by_id(id_val):  
    global size, data_version_one  # look at global size, data
    for i in range(size):  
        if data_version_one[i]['id'] == id_val:  # check if it matches
            return i  
    return -1  

def checking_capacity():  # checking to see if theres more room for more entries
    global capacity, data_version_one, size  # looking at globals
    if size < capacity:  # space = do nothing
        return
    new_capacity = capacity * 2  
    updated_data = [None] * new_capacity  # new list
    for i in range(size): 
        updated_data[i] = data_version_one[i]  
    data_version_one = updated_data  # replacing the old data with updated_data
    capacity = new_capacity  # update capacity

def add_entry_to_storage(id_val, score_val):  # to add an entry to storage
    global size, data_version_one  # looking at globals
    checking_capacity()  # calling the checker 
    data_version_one[size] = {'id': id_val, 'score': score_val}  # stored in dict
    size += 1  

def remove_entry_by_id(id_val):  # used to remove an entry from storage
    global size, data_version_one  # looking at globals
    idx = find_index_by_id(id_val)  
    if idx == -1:  
        return False  

    for i in range(idx, size - 1):
        data_version_one[i] = data_version_one[i + 1]  # shift left
    data_version_one[size - 1] = None  # clear last 
    size -= 1  # decrement 
    return True  

def update_by_pointer(idx, new_score):  
    global data_version_one  # looking at global data var
    if idx >= 0:
        data_version_one[idx]['score'] = new_score  

def update_by_reference(entry, new_score):  
    entry['score'] = new_score  

def boost_by_pointer(idx, delta):  # boost through pointer
    global data_version_one  # looking at globals
    if idx < 0:
        return
    temp = data_version_one[idx]['score'] + delta  # get the new value
    if temp < 0:  # set the lower bound
        temp = 0
    if temp > 1000000:  # set the upper bound
        temp = 1000000
    data_version_one[idx]['score'] = temp  

def boost_by_reference(entry, delta):  # boost through reference
    temp = entry['score'] + delta  
    if temp < 0:  # lower bound
        temp = 0
    if temp > 1000000:  # upper bound
        temp = 1000000
    entry['score'] = temp  


def print_menu(): 
    print("High Score Lab") 
    print("1. Add Entry") 
    print("2. View Entry")  
    print("3. Update Score")  
    print("4. Remove Entry") 
    print("5. Leaderboard (Top K)")  
    print("6. Stats")  
    print("7. Pointer Tools") 
    print("8. Exit")  
    print("Enter your choice: ", end='')  


def main():  
    global size, capacity, data_version_one  # get globals
    while True:  
        print_menu()  
        try:
            line = input()  
        except EOFError:
            return  # added just incase of console input, i had issues with the last assignment with it
        ok, val = try_to_parse_int(line)  
        if not ok:  
            print("Invalid input.")  
            print()  
            continue  
        choice = val  
        if choice < 1 or choice > 8:  
            print("Invalid choice.")  
            print()  
            continue  

        if choice == 8:  
            print("Goodbye!")  
            return  

        if choice == 1:  
            
            while True:
                print("Enter ID: ", end='')  
                try:
                    line = input()  
                except EOFError:
                    return
                ok, id_value = try_to_parse_int(line)  
                if not ok:  
                    print("Invalid input.")  
                    continue
                if id_value <= 0:  
                    print("Error: ID must be greater than 0.")  
                    continue  
                if find_index_by_id(id_value) != -1:  
                    print("Error: ID already exists.")  
                    continue  
                break  

            
            while True:
                print("Enter Score: ", end='')
                try:
                    line = input()
                except EOFError:
                    return
                ok, sval = try_to_parse_int(line)
                if not ok:
                    print("Invalid input.")  
                    continue  
                if sval < 0 or sval > 1000000:
                    print("Error: Score out of range.")  
                    continue  
                break  

            add_entry_to_storage(id_value, sval)  
            print("Entry added!")  
            print()  
            continue  

        if choice == 2:  
            print("Enter ID: ", end='')  
            try:
                line = input()
            except EOFError:
                return
            ok, id_value = try_to_parse_int(line)  
            if not ok:
                print("Invalid input.")  
                print()  
                continue  
            idx = find_index_by_id(id_value)  
            if idx == -1:
                print("Error: ID not found.")  
                print()  
                continue  
            print(f"ID: {data_version_one[idx]['id']} | Score: {data_version_one[idx]['score']}")  
            print()  
            continue  

        if choice == 3:  
            print("Enter ID: ", end='')  
            try:
                line = input()
            except EOFError:
                return
            ok, id_value = try_to_parse_int(line)  
            if not ok:
                print("Invalid input.")  
                print()  
                continue

            print("Enter new Score: ", end='')  
            try:
                line = input()
            except EOFError:
                return
            ok, newscore = try_to_parse_int(line)  
            if not ok:
                print("Invalid input.")  
                print()  
                continue

           
            while True:
                print("Method (1=pointer, 2=reference): ", end='')  
                try:
                    line = input()
                except EOFError:
                    return
                ok, method = try_to_parse_int(line)  
                if not ok or (method != 1 and method != 2):  
                    print("Invalid choice.")  
                    continue  
                break  

            idx = find_index_by_id(id_value)  # check if exists
            if idx == -1:
                print("Error: ID not found.")  
                print()  
                continue

            if newscore < 0 or newscore > 1000000:  
                print("Error: Score out of range.")  
                print()  
                continue


            if method == 1:
                update_by_pointer(idx, newscore)  # pointer update
            else:
                update_by_reference(data_version_one[idx], newscore)  # reference update

            print("Updated.")  
            print()  
            continue  

        if choice == 4:  
            
            while True:
                print("Enter ID: ", end='')  
                try:
                    line = input()
                except EOFError:
                    return
                ok, id_value = try_to_parse_int(line)  
                if not ok:
                    print("Invalid input.")  
                    continue  
                if not remove_entry_by_id(id_value):  
                    print("Error: ID not found.")  
                    continue  
                print("Removed.")  
                print()  
                break  
            continue  

        if choice == 5:
            print("Enter K: ", end='')  
            try:
                line = input()
            except EOFError:
                return
            ok, k = try_to_parse_int(line) 
            if not ok:
                print("Invalid input.")  
                print()  
                continue  
            if k <= 0:
                print("Error: K must be greater than 0.")  
                print()  
                continue  
            if size == 0:  
                print()  
                continue  

            temp = []  
            for i in range(size):
                temp.append({'id': data_version_one[i]['id'], 'score': data_version_one[i]['score']})  # copy
            
            temp.sort(key=lambda e: (-e['score'], e['id']))  
            to_show = min(len(temp), k)  
            for i in range(to_show):
                print(f"{i+1}. ID: {temp[i]['id']} | Score: {temp[i]['score']}")  
            print()  
            continue  

        if choice == 6:  
            total = 0
            for i in range(size):
                total += data_version_one[i]['score']
            avg = 0.0  
            if size > 0:
                avg = total / size  
            
            print(f"Entries: {size}")  
            print(f"Total score: {total}")  
            print("Average score: {0:.2f}".format(avg))  
            print(f"Capacity: {capacity}")  
            print()  
            continue  

        if choice == 7:  
            while True:
                print("a) Boost by pointer")  
                print("b) Boost by reference")  
                print("c) Back")  
                print("Enter choice: ", end='')  
                try:
                    line = input()
                except EOFError:
                    return
                if line not in ('a', 'b', 'c'):  
                    print("Invalid choice.")  
                    continue  
                if line == 'c':  
                    print()  
                    break  
                print("Enter ID: ", end='')  
                try:
                    line = input()
                except EOFError:
                    return
                ok, id_value = try_to_parse_int(line)  
                if not ok:
                    print("Invalid input.")  
                    continue  
                idx = find_index_by_id(id_value)  
                if idx == -1:
                    print("Error: ID not found.")  
                    continue  
                print("Enter delta: ", end='')  
                try:
                    line = input()
                except EOFError:
                    return
                ok, delta = try_to_parse_int(line)  
                if not ok:
                    print("Invalid input.")
                    continue  
                if line == 'a':  
                    boost_by_pointer(idx, delta) # call boost
                    print("Boosted.")
                    continue  
                else:  
                    boost_by_reference(data_version_one[idx], delta)  # call reference
                    print("Boosted.")  
                    continue  

if __name__ == "__main__":
    main()  
